src/actions/mod.rs

```diff
@@ -717,43 +717,51 @@ impl ActionHandler {
     }
 
     pub fn on_change_config<O: Output>(&self, params: DidChangeConfigurationParams, out: O) {
-        trace!("config change: {:?}", params.settings);
-        if let Some(config) = params.settings.get("rust") {
-            if let Ok(new_config) = serde_json::from_value(config.clone()): Result<Config, _> {
-                let unstable_features = new_config.unstable_features;
-
-                {
-                    let mut config = self.config.lock().unwrap();
-                    *config = new_config;
-                }
-                // We do a clean build so that if we've changed any relevant options
-                // for Cargo, we'll notice them. But if nothing relevant changes
-                // then we don't do unnecessary building (i.e., we don't delete
-                // artifacts on disk).
-                self.build_current_project(BuildPriority::Cargo, out.clone());
-
-                const RANGE_FORMATTING_ID: &'static str = "rls-range-formatting";
-                // FIXME should handle the response
-                if unstable_features {
-                    let output = serde_json::to_string(
-                        &RequestMessage::new(out.provide_id(),
-                                             NOTIFICATION__RegisterCapability.to_owned(),
-                                             RegistrationParams { registrations: vec![Registration { id: RANGE_FORMATTING_ID.to_owned(), method: REQUEST__RangeFormatting.to_owned(), register_options: serde_json::Value::Null }] })
-                    ).unwrap();
-                    out.response(output);
-                } else {
-                    let output = serde_json::to_string(
-                        &RequestMessage::new(out.provide_id(),
-                                             NOTIFICATION__UnregisterCapability.to_owned(),
-                                             UnregistrationParams { unregisterations: vec![Unregistration { id: RANGE_FORMATTING_ID.to_owned(), method: REQUEST__RangeFormatting.to_owned() }] })
-                    ).unwrap();
-                    out.response(output);
-                }
+        use serde::Deserialize;
+        use serde::de::Error;
 
+        trace!("config change: {:?}", params.settings);
+        let config = params.settings.get("rust")
+                         .ok_or(serde_json::Error::missing_field("rust"))
+                         .and_then(|value| Config::deserialize(value));
```

Comment:
Asked over at #serde and changed to `Type::serialize`, since it probably allows to deserialize from ref directly instead of cloning the Value?